protocol/xmrtaker/swap_state.go

```diff
@@ -154,13 +123,100 @@ func newSwapState(
 		exchangeRate,
 		ethAsset,
 		stage,
-		walletScanHeight,
+		moneroStartNumber,
 		statusCh,
 	)
 	if err = b.SwapManager().AddSwap(info); err != nil {
 		return nil, err
 	}
 
+	defer func() {
+		statusCh <- stage
+	}()
+	return newSwapState(
+		b,
+		transferBack,
+		info,
+		ethHeader.Number,
+		moneroStartNumber,
+	)
+}
+
+func newSwapStateFromOngoing(
+	b backend.Backend,
+	info *pswap.Info,
+	transferBack bool,
+	ethSwapInfo *db.EthereumSwapInfo,
+	sk *mcrypto.PrivateKeyPair,
+) (*swapState, error) {
+	if info.Status != types.ETHLocked && info.Status != types.ContractReady {
+		return nil, errInvalidStageForRecovery
+	}
+
+	makerSk, makerVk, err := b.RecoveryDB().GetXMRMakerSwapKeys(info.ID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get xmrmaker swap keys from db: %w", err)
+	}
+
+	s, err := newSwapState(
+		b,
+		transferBack,
+		info,
+		ethSwapInfo.StartNumber,
+		info.MoneroStartHeight,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if b.ContractAddr() != ethSwapInfo.ContractAddress {
+		return nil, errContractAddrMismatch(ethSwapInfo.ContractAddress.String())
```

Comment:
In the longer term, who is going to be specifying the contract address? Will it be Bob (the maker), so that he can control what trusted forwarder is used? Forgetting about what is appropriate for this pull request, I'm wondering if this is a scenario that we'll need to be able to handle later on.

Reply:
so far it's Alice who controls what contract is used, since she sends the address when notifying Bob of `newSwap` being executed. 

in terms of the trusted forwarder, Bob will need to verify the forwarder bytecode if he decides to use a relayer